stablehlo/transforms/StablehloProbeInstrumentation.cpp

```diff
@@ -0,0 +1,113 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+   Copyright 2023 The StableHLO Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include <string>
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOPROBEINSTRUMENTATIONPASS
```

Comment:
Lets think about pass name / command line flag.

The relation should be:

```
StablehloDoSomethingPass -->  --stablehlo-do-something
```

In this case it would be

```
StablehloProbeInstrumentationPass --> --stablehlo-probe-instrumentation
```

And I don't think that's a proper flag name so perhaps we should rename this pass as a verb.

```
StablehloInsertProbesPass --> --stablehlp-insert-probes
StablehloInstrumentProbePass --> --stablehlo-instrument-probe
StablehloInstrumentWithProbePass --> --stablehlo-instrument-with-probe
StablehloProbeInstrumentModulePass --> --stablehlo-probe-instrument-module
```

Let me know if you have any other ideas for this, but I think we should land on something more declarative

Reply:
I like the idea of dropping `Instrumentation` as it does sound redundant given `Probe`, and naming the pass `StablehloProbePass`. Then this could fall in line with existing pass nomenclature to yield `--stablehlo-probe` as @GleasonK mentioned earlier.

Reply:
Also, I may open another PR to rename existing `probeInstrumentation` nomenclature due to other recent PRs, to better unify the pass name here with `InterpreterConfiguration` settings.

Reply:
Re-opening, sorry to nitpick on names, I just think if a typical person saw `-stablehlo-probe` used in a file they wouldn't be able to describe whats going on.

> I think in this context probe and instrument mean the same thing

In this context one is a verb one is a noun. All of our passes follow a pattern of "noun-verb-noun", ex: `stablehlo-canonicalize-dynamism`, `stablehlo-refine-shapes`, `stablehlo-legalize-to-*`, ... The only outlier is `vhlo-to-version` which is also probably a bad name that should be revisited, `vhlo-legalize-to-version`.

With that said, I'd suggest either use `probe` as a noun referring to probe ob `ProbeOp`:

- `StablehloInjectProbeOpPass, --stablehlo-inject-probe`
- `StablehloInstrumentWithProbePass, --stablehlo-instrument-with-probe`

Or a verb, meaning "to probe the op"

- `StablehloProbeModule --stablehlo-probe-module`
- `StablehloProbeOps --stablehlo-probe-ops`

With that said, I prefer `StablehloInstrumentWithProbe` or `StablehloInjectProbeOpPass`. We don't tend to get super concise on pass names, the same is true of MHLO transformation pass names like `HloLegalizeShapeOpsToStandardPass -hlo-legalize-shapeops-to-standard`

Reply:
I prefer the `StablehloInstrumentWithProbe` over `StablehloInjectProbeOpPass` nomenclature; updated the pass name and flag.